---
title: "ASG_Project"
author: "Thomas Dechatre and Zacharie Ménétrier"
date: "January 13, 2018"
output: html_document
references:
- id: ref
  author:
  - family: Labadorf
    given: A
  - family: Lagomarsino
    given: V
  - family: Latourelle
    given: JC
  title: mRNA-Seq Expression profiling of human post-mortem BA9 brain tissue for Huntington''s Disease and neurologically normal individuals
  type: article-journal
  issued:
  year: 2015
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Huntigton's disease and gene expression

"Huntington's Disease (HD) is a devastating neurodegenerative disorder that is caused by an expanded CAG trinucleotide repeat in the Huntingtin (HTT) gene." [@ref] 

The study contains 49 control samples and 20 samples with Huntigton's disease, for a total of more than 58000 variables accounted.
In the first part of this project we will apply a differential expression gene analysis in order to discover some genes statistically implied in Huntington's disease.
In the second part, two machine learning methods will be evaluated to classify healthy control from tested gene expression. We will also try to demonstrate the pertinence of some feature selections methods.

### Setting a path on your computer

In order to download some files, a path must be specified for the rest of the project to work correctly.

```{r}
setwd("/home/zac/Desktop/ExtraMarker/")
```

## Differential expression analysis

In this part of the study we will apply a differential expression analysis to our dataset. We will first get some statistics about the data before normalizing the counts and try to apply a functional enrichment methods to get some genes implied in Huntigton's disease.

### Downloading the study

The study results can be found on the recount database. The id of this study under the recount database is **SRP007825**. We will use some methods of the recount R package in order to facilitate the loading of the count table. The recount package can be found under the bioconductor software set.

```{r message=FALSE, warning=FALSE, paged.print=FALSE}
# Installing the recount R package
source("https://bioconductor.org/biocLite.R")
if(!require("recount")){
  biocLite("recount")
}
library(recount)
# Downloading the study files
project.id <- 'SRP051844'
if(!file.exists(file.path(project.id, 'rse_gene.Rdata'))){
  download_study(project.id)
}
```

The two things we will need out of the study is the count table of RNA reads and the type (control or disease) for each samples.

```{r load data}
# Load the rse data
load(file.path(project.id, 'rse_gene.Rdata'))
# Load the characteristics of each sample
geochar <- lapply(split(colData(rse_gene), seq_len(nrow(colData(rse_gene)))), recount::geo_characteristics)
# Get the phenotypes of interest
phenotypes <- unlist(as.factor(unlist(lapply(geochar, "[", 2))))
# Create the DESeq2 object
rse <- scale_counts(rse_gene)
# The count table is created as a matrix of count
count.table <- as.matrix(assay(rse))
```

### First glance at the data

We will now try to see some interesting statistics about the loaded tables.
The first thing we would like to know is the number of read for each sample.

```{r barplot}
# A color is assigned to each condition
col.phenotypes <- c("Neurologically normal"="cadetblue", "Huntington's Disease"="tomato")
colors <- col.phenotypes[as.vector(phenotypes)]
# Creation of the bar plot
barplot(colSums(count.table)/1000000, 
        main="Total number of reads per sample (million)",
        col=colors,
        border = "white",
        #        names.arg = "", 
        las=1,  horiz=TRUE,
        ylab="Samples", cex.names=0.2,
        xlab="Million counts")
```

Here we see that the number of reads per sample is rather consistent among samples.

```{r Computation for the raw counts}
moyenneRaw <- colMeans(count.table)
medianeRaw <- colMedians(count.table)
maxRaw <- colMaxs(count.table)
minRaw <- colMins(count.table)
quantileRaw <- colQuantiles(count.table)
zeroRaw <- colCounts(count.table, value = 0)
head(data.frame(moyenne= moyenneRaw, mediane = medianeRaw, quartile0 = quantileRaw[,1], quartile25 = quantileRaw[,2], quartile50 = quantileRaw[,3], quartile75 = quantileRaw[,4], quartile100 = quantileRaw[,5] ,max = maxRaw, min = minRaw, NombreZeros = zeroRaw, row.names = colnames(count.table)))
```

Summary of the statistics per samples of the raw counts into the dataframe named "dataRawStatistics".
```{r Computation the log2-transformed counts}
log2count.table <- log2(count.table+1)
moyenneLog <- colMeans(log2count.table)
medianeLog <- colMedians(log2count.table)
maxLog <- colMaxs(log2count.table)
minLog <- colMins(log2count.table)
quantileLog <- colQuantiles(log2count.table)
zeroLog <- colCounts(log2count.table, value = 0)
head(data.frame(moyenne= moyenneLog, mediane = medianeLog, quartile0 = quantileLog[,1], quartile25 = quantileLog[,2], quartile50 = quantileLog[,3], quartile75 = quantileLog[,4], quartile100 = quantileLog[,5] ,max = maxRaw, min = minLog,NombreZeros = zeroLog, row.names = colnames(log2count.table)))
```
Summary of the statitics of the log2-transformed counts into the dataframe named "dataLogStatistics".

It would be important to visualize the distribution of number of reads for gene expression. However most genes show near zero counts, this is why the log2 transformed table is used to show such distribution.

```{r histogram}
# Creating an histogram of log2 counts per gene
hist(as.matrix(log2(count.table + 1)), col="slategray", border="white",
     main="Log2-transformed counts per gene", xlab="log2(counts+1)", ylab="Number of genes", 
     las=1, cex.axis=0.7)
```

A box plot of the samples could determine if a normalisation step is necessary.

```{r}
boxplot(log2(count.table+ 1), col=colors, pch=".", 
        horizontal=TRUE, cex.axis=0.5,
        las=1, ylab="Samples", xlab="log2(Counts +1)")
```

Here we see the expression of genes for each sample show no strong differences, the normalisation step should only slightly change our dataste values.

### Elimination of undected genes

```{r elimination of undected genes}
reduced.count.table <- count.table[rowSums(count.table) > 0,]
nrow(count.table)
nrow(reduced.count.table)
```

Eliminating undected genes has removed approximately 7000 genes from the data.

### Differential expression analysis

To be able to perform a differential expression analysis we will load the DESeq2 R package which contains functions that will help us.

```{r}
if(!require("DESeq2")){
  biocLite("DESeq2")
}
library(DESeq2)
```

After that, a DESeqDataSet object must be created from our count table.

```{r}
# Creation of a data frame that will contains the informations of the study
colData <- data.frame(colnames(reduced.count.table))
colData <- cbind(colData, phenotypes)
colnames(colData) <- c("Sample", "Phenotype")
# The column names of the count table are set to null to avoid an error
colnames(reduced.count.table) <- NULL
dds <- DESeqDataSetFromMatrix(countData = reduced.count.table, colData = colData, design = ~ Phenotype)
# The column names of the count table are reset
colnames(reduced.count.table) <- colnames(count.table)
# Estimation of the size factor for each sample
dds.norm <-  estimateSizeFactors(dds)
# Creation of a barplot that show the size factor for each sample
barplot(sizeFactors(dds.norm), 
        main="Size factor for each sample",
        col=colors,
        border = "white",
        #        names.arg = "", 
        las=1,  horiz=TRUE,
        ylab="Samples", cex.names=0.2,
        xlab="Size factors")
```

Here we see the size factors for each sample is very close to 1. That means normalisation between samples will not greatly change the number of counts.

We will now see some plots of the data before and after the normalisation to see what has changed.

```{r}
## Checking the normalization
par(mfrow=c(1,2),cex.lab=0.7)
boxplot(log2(counts(dds.norm) + 1),  col=colors, cex.axis=0.7, 
        las=1, xlab="log2(counts)", horizontal=TRUE, main="Raw counts")
boxplot(log2(counts(dds.norm, normalized=TRUE) + 1),  col=colors, cex.axis=0.7, 
        las=1, xlab="log2(normalized counts)", horizontal=TRUE, main="Normalized counts") 
par(mfrow=c(1,1),cex.lab=1)
```

The mean and the variance are not linearly related.

```{r}
## Computing mean and variance
norm.counts <- counts(dds.norm, normalized=TRUE)
mean.counts <- rowMeans(norm.counts)
variance.counts <- apply(norm.counts, 1, var)
## Mean and variance relationship
mean.var.col <- densCols(x=log2(mean.counts), y=log2(variance.counts))
plot(x=log2(mean.counts), y=log2(variance.counts), pch=16, cex=0.5, 
     col=mean.var.col, main="Mean-variance relationship",
     xlab="Mean log2(normalized counts) per gene",
     ylab="Variance of log2(normalized counts)",
     panel.first = grid())
abline(a=0, b=1, col="brown")
```

```{r}
## Performing estimation of dispersion parameter
dds.disp <- estimateDispersions(dds.norm)
## A diagnostic plot which
## shows the mean of normalized counts (x axis)
## and dispersion estimate for each genes
plotDispEsts(dds.disp)
#pvalues
alpha <- 0.0001
wald.test <- nbinomWaldTest(dds.disp)
res.DESeq2 <- results(wald.test, alpha=alpha, pAdjustMethod="BH")
res.DESeq2 <- res.DESeq2[order(res.DESeq2$padj),]
head(res.DESeq2)
## Draw an histogram of the p-values
hist(res.DESeq2$padj, breaks=20, col="grey", main="DESeq2 p-value distribution", xlab="DESeq2 P-value", ylab="Number of genes")
#volcano  plot
alpha <- 0.01 # Threshold on the adjusted p-value
cols <- densCols(res.DESeq2$log2FoldChange, -log10(res.DESeq2$pvalue))
plot(res.DESeq2$log2FoldChange, -log10(res.DESeq2$padj), col=cols, panel.first=grid(),
     main="Volcano plot", xlab="Effect size: log2(fold-change)", ylab="-log10(adjusted p-value)",
     pch=20, cex=0.6)
abline(v=0)
abline(v=c(-1,1), col="brown")
abline(h=-log10(alpha), col="brown")
gn.selected <- abs(res.DESeq2$log2FoldChange) > 2 & res.DESeq2$padj < alpha 
## Draw a MA plot.
## Genes with adjusted p-values below 1% are shown
plotMA(res.DESeq2, colNonSig = "blue")
abline(h=c(-1:1), col="red")
```

```{r}
if(!require("gplots")){
  install.packages("gplots")
}
library("gplots")
```

```{r}
## We select gene names based on FDR (1%)
gene.kept <- rownames(res.DESeq2)[res.DESeq2$padj <= alpha & !is.na(res.DESeq2$padj)]
## We retrieve the normalized counts for gene of interest
count.table.kept <- log2(count.table + 1)[gene.kept, ]
## Perform the hierarchical clustering with
## A distance based on Pearson-correlation coefficient
## and average linkage clustering as agglomeration criteria
heatmap.2(as.matrix(count.table.kept), 
          scale="row", 
          hclust=function(x) hclust(x,method="average"), 
          distfun=function(x) as.dist((1-cor(t(x)))/2), 
          trace="none", 
          density="none", 
          labRow="",
          cexCol=0.7)
```

## Supervised classification

In this part of the study we will compare two machine learning models. We will demonstrate the importance of dimensionality and feature selection.

To work with machine learning and feature selection, a data frame should be created that contains all the variables normalized and transformed with a logarithmic function.

```{r}
all.data <- data.frame(t(log2(counts(dds.norm, normalized=TRUE) + 1)))
```

### Feature selection

Our data frame contains more than 51000 genes which are the descriptive variables of our set. For a lot of machine learning models, it is too much and may lead to overfitting, and so it is important to try to reduce the number of features. We will try to apply two feature selection methods, ANOVA and PCA to reduce the number of descriptive variables.

#### Filter by variance

A naive feature selection would be to assume that gene variance is related to differential expression in both conditions.
We will try to order the genes by variance before visualizing if a clear separation is effective.

```{r}
var.per.genes <- colVars(as.matrix(all.data))
var.order <- order(var.per.genes, decreasing = TRUE)
# Plot of the two genes with the highest variance
plot(all.data[,var.order[1]], all.data[,var.order[2]], col = colors, pch = 19,
     main = "Plot of the two genes with the highest variances", xlab = colnames(all.data)[var.order[1]],
     ylab = colnames(all.data)[var.order[2]])
```



